drivers/i2c/i2c_numaker.c

```diff
@@ -0,0 +1,808 @@
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(i2c_numaker, CONFIG_I2C_LOG_LEVEL);
+
+#include "i2c-priv.h"
+#include <soc.h>
+#include <NuMicro.h>
+
+/* i2c Master Mode Status */
+#define  M_START               0x08  /* Start */
+#define  M_REPEAT_START        0x10  /* Master Repeat Start */
+#define  M_TRAN_ADDR_ACK       0x18  /* Master Transmit Address ACK */
+#define  M_TRAN_ADDR_NACK      0x20  /* Master Transmit Address NACK */
+#define  M_TRAN_DATA_ACK       0x28  /* Master Transmit Data ACK */
+#define  M_TRAN_DATA_NACK      0x30  /* Master Transmit Data NACK */
+#define  M_ARB_LOST            0x38  /* Master Arbitration Los */
+#define  M_RECE_ADDR_ACK       0x40  /* Master Receive Address ACK */
+#define  M_RECE_ADDR_NACK      0x48  /* Master Receive Address NACK */
+#define  M_RECE_DATA_ACK       0x50  /* Master Receive Data ACK */
+#define  M_RECE_DATA_NACK      0x58  /* Master Receive Data NACK */
+#define  BUS_ERROR             0x00  /* Bus error */
+
+/* i2c Slave Mode Status */
+#define  S_REPEAT_START_STOP   0xA0  /* Slave Transmit Repeat Start or Stop */
+#define  S_TRAN_ADDR_ACK       0xA8  /* Slave Transmit Address ACK */
+#define  S_TRAN_DATA_ACK       0xB8  /* Slave Transmit Data ACK */
+#define  S_TRAN_DATA_NACK      0xC0  /* Slave Transmit Data NACK */
+#define  S_TRAN_LAST_DATA_ACK  0xC8  /* Slave Transmit Last Data ACK */
+#define  S_RECE_ADDR_ACK       0x60  /* Slave Receive Address ACK */
+#define  S_RECE_ARB_LOST       0x68  /* Slave Receive Arbitration Lost */
+#define  S_RECE_DATA_ACK       0x80  /* Slave Receive Data ACK */
+#define  S_RECE_DATA_NACK      0x88  /* Slave Receive Data NACK */
+
+/* i2c GC Mode Status */
+#define  GC_ADDR_ACK           0x70  /* GC mode Address ACK */
+#define  GC_ARB_LOST           0x78  /* GC mode Arbitration Lost */
+#define  GC_DATA_ACK           0x90  /* GC mode Data ACK */
+#define  GC_DATA_NACK          0x98  /* GC mode Data NACK */
+
+/* i2c Other Status */
+#define  ADDR_TRAN_ARB_LOST    0xB0  /* Address Transmit Arbitration Lost */
+#define  BUS_RELEASED          0xF8  /* Bus Released */
+
+/* Implementation notes
+ * 1. Support dual role
+ * 2. Support at most one slave at one time.
+ * 3. Support merging adjacent messages into one transaction.
+ * 4. Force I2C Stop to end the transfer
+ */
+
+/* Transfer control type
+ * Active transfer (master only):
+ * 1. Do I2C Start to start the transfer (thread)
+ * 2. I2C FSM (ISR)
+ * 3. Force I2C Stop to end the transfer (thread)
+ *
+ * Passive transfer (slave only):
+ * 1. Prepare callback (thread)
+ * 2. Do data transfer via above callback (ISR)
+ */
+
+static void m_numaker_i2c_master_xfer_msg_read_next_byte(const struct device *dev);
+static void m_numaker_i2c_master_xfer_msg_read_last_byte(const struct device *dev);
+static void m_numaker_i2c_master_xfer_msg_end(const struct device *dev);
+static void m_numaker_i2c_master_xfer_end(const struct device *dev, bool do_stop);
```

Comment:
could you move those functions up and avoid having to forward declare them? It'd also be nice to have `DT_INST_FOREACH_STATUS_OKAY` last

Reply:
of course, I can follow it.